provider/pkg/await/deployment.go

```diff
@@ -369,6 +376,45 @@ func (dia *deploymentInitAwaiter) await(
 	}
 }
 
+func (dia *deploymentInitAwaiter) waitUntilDeploymentControllerReconciles(deploymentEvents <-chan watch.Event, timeout <-chan time.Time) error {
+	for {
+		select {
+		case <-dia.config.ctx.Done():
+			return &cancellationError{
+				object:    dia.deployment,
+				subErrors: dia.errorMessages(),
+			}
+		case <-timeout:
+			return &timeoutError{
+				object:    dia.deployment,
+				subErrors: dia.errorMessages(),
+			}
+		case event := <-deploymentEvents:
+			deployment, isUnstructured := event.Object.(*unstructured.Unstructured)
+			if !isUnstructured {
+				logger.V(3).Infof("Deployment watch received unknown object type %T",
+					event.Object)
+				continue
+			}
+
+			// Do nothing if this is not the Deployment we're waiting for.
+			if deployment.GetName() != dia.deployment.GetName() {
+				continue
+			}
+
+			observedGeneration, found, err := unstructured.NestedInt64(deployment.Object, "status", "observedGeneration")
+			if err != nil || !found {
+				continue
+			}
+
+			if deployment.GetGeneration() == observedGeneration {
```

Comment:
Would `generation <= observedGeneration` be technically safer here? I don't know if that's actually observable in practice.

Reply:
We need the `observedGeneration` to match exactly the `generation` of the deployment object. `generation <= observedGeneration` is currently why this bug exists.

Reply:
I too stared at the comparison for a few minutes.  Rollout code will sometimes have a 'planned' generation that they got from the call to `Apply`, and await an observed generation that is greater-or-equal to the planned generation.  This code is not doing that; it is simply comparing generations within a given object and has no planned generation.